@ghalliday There are tons of errors like this:
```
      1. Test: HeadingExample.ecl
      eclcc returns with:
        Error (1001): Command return code non-zero.
         retCode: 2
        /home/centos/smoketest/PR-15331/HPCC-Platform/testing/regress/ecl/HeadingExample.ecl(1,0): error C0003: assert(name) failed - file: hqlexpr.cpp, line 1334
        1 error, 0 warning

```

--- a/common/fileview2/fvtransform.cpp
+++ b/common/fileview2/fvtransform.cpp
@@ -297,7 +297,7 @@ void ViewTransformerRegistry::addPlugins(const char * name)
     Owned<IErrorReceiver> errorReporter = createThrowingErrorReceiver();
     EclRepositoryManager collection;
     collection.addSharedSourceFileEclRepository(errorReporter, name, ESFallowplugins, 0, false);
-    dataServer.setown(collection.createCompoundRepository());
+    dataServer.setown(collection.createPackage(nullptr));
 
     NullStatisticTarget nullStats;
     HqlScopeArray scopes;

--- a/common/fileview2/fvtransform.ipp
+++ b/common/fileview2/fvtransform.ipp
@@ -228,7 +228,7 @@ protected:
     ViewFieldTransformerArray transformers;
     CIArrayOf<ViewServiceEntry> plugins;
     Owned<SafePluginMap> loadedPlugins;
-    Owned<IEclRepository> dataServer;
+    Owned<IEclPackage> dataServer;
     SimplePluginCtx pluginCtx;
 };
 

--- a/ecl/eclcc/eclcc.cpp
+++ b/ecl/eclcc/eclcc.cpp
@@ -211,7 +211,7 @@ public:
     Linked<IFile> inputFile;
     Linked<IPropertyTree> archive;
     Linked<IWorkUnit> wu;
-    Owned<IEclRepository> dataServer;  // A member which can be cleared after parsing the query
+    Owned<IEclPackage> dataServer;  // A member which can be cleared after parsing the query
     OwnedHqlExpr query;  // parsed query - cleared when generating to free memory
     StringAttr eclVersion;
     const char * outputFilename;
@@ -300,10 +300,11 @@ protected:
     void outputXmlToOutputFile(EclCompileInstance & instance, IPropertyTree * xml);
     void processSingleQuery(EclCompileInstance & instance,
                                IFileContents * queryContents,
-                               const char * queryAttributePath);
+                               const char * queryAttributePath,
+                               IEclPackage * mainPackage);
     void processXmlFile(EclCompileInstance & instance, const char *archiveXML);
     void processFile(EclCompileInstance & info);
-    void processReference(EclCompileInstance & instance, const char * queryAttributePath);
+    void processReference(EclCompileInstance & instance, const char * queryAttributePath, const char * queryAttributePackage);
     void processBatchFiles();
     void processDefinitions(EclRepositoryManager & target);
     void reportCompileErrors(IErrorReceiver & errorProcessor, const char * processName);
@@ -336,7 +337,8 @@ protected:
     StringAttr optIniFilename;
     StringAttr optOutputDirectory;
     StringAttr optOutputFilename;
-    StringAttr optQueryRepositoryReference;
+    StringAttr optQueryMainAttribute;
+    StringAttr optQueryMainPackage;
     StringAttr optComponentName;
     StringAttr optDFS;
     StringAttr optCluster;
@@ -605,8 +607,11 @@ void EclCC::loadManifestOptions()
             optLegacyWhen = ecl->getPropBool("@legacyWhen", optLegacy);
         }
 
-        if (!optQueryRepositoryReference && ecl->hasProp("@main"))
-            optQueryRepositoryReference.set(ecl->queryProp("@main"));
+        if (!optQueryMainAttribute && ecl->hasProp("@main"))
+        {
+            optQueryMainAttribute.set(ecl->queryProp("@main"));
+            optQueryMainPackage.set(ecl->queryProp("@package"));
+        }
 
         if (ecl->hasProp("@targetPlatform"))
             setTargetPlatformOption(ecl->queryProp("@targetPlatform"), optTargetClusterType);
@@ -1164,7 +1169,8 @@ void EclCC::evaluateResult(EclCompileInstance & instance)
 
 void EclCC::processSingleQuery(EclCompileInstance & instance,
                                IFileContents * queryContents,
-                               const char * queryAttributePath)
+                               const char * queryAttributePath,
+                               IEclPackage * mainPackage)
 {
 #ifdef TEST_LEGACY_DEPENDENCY_CODE
     setLegacyEclSemantics(instance.legacyImportMode, instance.legacyWhenMode);
@@ -1277,8 +1283,10 @@ void EclCC::processSingleQuery(EclCompileInstance & instance,
         instance.archive->setPropBool("@legacyWhen", instance.legacyWhen);
         if (withinRepository)
         {
+            IEclPackage * package = mainPackage ? mainPackage : instance.dataServer.get();
             instance.archive->setProp("Query", "");
             instance.archive->setProp("Query/@attributePath", queryAttributePath);
+            instance.archive->setProp("Query/@package", package->queryPackageName());
         }
     }
 
@@ -1373,7 +1381,7 @@ void EclCC::processSingleQuery(EclCompileInstance & instance,
 
             if (withinRepository)
             {
-                instance.query.setown(getResolveAttributeFullPath(queryAttributePath, LSFpublic, ctx));
+                instance.query.setown(getResolveAttributeFullPath(queryAttributePath, LSFpublic, ctx, mainPackage));
                 if (!instance.query && !syntaxChecking && (errorProcessor.errCount() == prevErrs))
                 {
                     StringBuffer msg;
@@ -1631,9 +1639,14 @@ void EclCC::processXmlFile(EclCompileInstance & instance, const char *archiveXML
 
     const char * queryText = archiveTree->queryProp("Query");
     const char * queryAttributePath = archiveTree->queryProp("Query/@attributePath");
+    const char * queryAttributePackage = archiveTree->queryProp("Query/@package");
+
     //Takes precedence over an entry in the archive - so you can submit parts of an archive.
-    if (optQueryRepositoryReference)
-        queryAttributePath = optQueryRepositoryReference;
+    if (optQueryMainAttribute)
+    {
+        queryAttributePath = optQueryMainAttribute;
+        queryAttributePackage = optQueryMainPackage;
+    }
 
     //The legacy mode (if specified) in the archive takes precedence - it needs to match to compile.
     instance.legacyImport = archiveTree->getPropBool("@legacyMode", instance.legacyImport);
@@ -1649,21 +1662,11 @@ void EclCC::processXmlFile(EclCompileInstance & instance, const char *archiveXML
     if (optCheckEclVersion)
         instance.checkEclVersionCompatible();
 
-    Owned<IEclSourceCollection> archiveCollection;
-    if (archiveTree->getPropBool("@testRemoteInterface", false))
-    {
-        //This code is purely here for regression testing some of the classes used in the enterprise version.
-        Owned<IXmlEclRepository> xmlRepository = createArchiveXmlEclRepository(archiveTree);
-        archiveCollection.setown(createRemoteXmlEclCollection(NULL, *xmlRepository, NULL, false));
-        archiveCollection->checkCacheValid();
-    }
-    else
-        archiveCollection.setown(createArchiveEclCollection(archiveTree));
-
     EclRepositoryManager localRepositoryManager;
     processDefinitions(localRepositoryManager);
     localRepositoryManager.inherit(repositoryManager); // Definitions, plugins, std library etc.
     Owned<IFileContents> contents;
+    IEclPackage * mainPackage = nullptr;
     StringBuffer fullPath; // Here so it doesn't get freed when leaving the else block
 
     if (queryText || queryAttributePath)
@@ -1695,15 +1698,16 @@ void EclCC::processXmlFile(EclCompileInstance & instance, const char *archiveXML
         localRepositoryManager.addSingleDefinitionEclRepository(syntaxCheckModule, syntaxCheckAttribute, contents, true);
     }
 
-    localRepositoryManager.addRepository(archiveCollection, nullptr, true);
+    localRepositoryManager.processArchive(archiveTree);
 
-    instance.dataServer.setown(localRepositoryManager.createCompoundRepository());
+    instance.dataServer.setown(localRepositoryManager.createPackage(nullptr));
+    if (queryAttributePackage)
+        mainPackage = localRepositoryManager.resolveDependentRepository(nullptr, queryAttributePackage, false);
 
     //Ensure classes are not linked by anything else
-    archiveCollection.clear();
     localRepositoryManager.kill();  // help ensure non-shared repositories are freed as soon as possible
 
-    processSingleQuery(instance, contents, queryAttributePath);
+    processSingleQuery(instance, contents, queryAttributePath, mainPackage);
 }
 
 
@@ -1747,7 +1751,7 @@ void EclCC::processFile(EclCompileInstance & instance)
 
     //On a system with userECL not allowed, all compilations must be from checked-in code that has been
     //deployed to the eclcc machine via other means (typically via a version-control system)
-    if (!allowAccess("userECL", false) && (!optQueryRepositoryReference || queryText->length()))
+    if (!allowAccess("userECL", false) && (!optQueryMainAttribute || queryText->length()))
     {
         instance.queryErrorProcessor().reportError(HQLERR_UserCodeNotAllowed, HQLERR_UserCodeNotAllowed_Text, NULL, 1, 0, 0);
     }
@@ -1759,13 +1763,15 @@ void EclCC::processFile(EclCompileInstance & instance)
     else
     {
         StringBuffer attributePath;
+        const char * attributePackage = nullptr;
         bool withinRepository = false;
 
         //Specifying --main indicates that the query text (if present) replaces that definition
-        if (optQueryRepositoryReference)
+        if (optQueryMainAttribute)
         {
             withinRepository = true;
-            attributePath.clear().append(optQueryRepositoryReference);
+            attributePath.clear().append(optQueryMainAttribute);
+            attributePackage = optQueryMainPackage;
         }
         else
         {
@@ -1783,7 +1789,7 @@ void EclCC::processFile(EclCompileInstance & instance)
         if (withinRepository)
         {
             //-main only overrides the definition if the query is non-empty.  Otherwise use the existing text.
-            if (!optQueryRepositoryReference || queryText->length())
+            if (!optQueryMainAttribute || queryText->length())
             {
                 Owned<IEclSourceCollection> inputFileCollection = createSingleDefinitionEclCollection(attributePath, queryText);
                 localRepositoryManager.addRepository(inputFileCollection, nullptr, true);
@@ -1813,11 +1819,19 @@ void EclCC::processFile(EclCompileInstance & instance)
             }
         }
 
-        localRepositoryManager.addQuerySourceFileEclRepository(&instance.queryErrorProcessor(), querySearchPath.str(), 0, 0);
+        if (attributePackage)
+        {
+            //If attribute package is specified, resolve that package as the source for the query, and pass null to processSingleQuery
+            instance.dataServer.setown(localRepositoryManager.resolveDependentRepository(nullptr, attributePackage, false));
+        }
+        else
+        {
+            localRepositoryManager.addQuerySourceFileEclRepository(&instance.queryErrorProcessor(), querySearchPath.str(), 0, 0);
 
-        instance.dataServer.setown(localRepositoryManager.createCompoundRepository());
+            instance.dataServer.setown(localRepositoryManager.createPackage(nullptr));
+        }
         localRepositoryManager.kill();  // help ensure non-shared repositories are freed as soon as possible
-        processSingleQuery(instance, queryText, attributePath.str());
+        processSingleQuery(instance, queryText, attributePath.str(), nullptr);
     }
 
     if (!instance.reportErrorSummary() || instance.archive || (optGenerateMeta && instance.generatedMeta))
@@ -1980,7 +1994,7 @@ void EclCC::generateOutput(EclCompileInstance & instance)
 }
 
 
-void EclCC::processReference(EclCompileInstance & instance, const char * queryAttributePath)
+void EclCC::processReference(EclCompileInstance & instance, const char * queryAttributePath, const char * queryAttributePackage)
 {
     const char * outputFilename = instance.outputFilename;
 
@@ -1993,10 +2007,29 @@ void EclCC::processReference(EclCompileInstance & instance, const char * queryAt
     localRepositoryManager.inherit(repositoryManager);
     if (!optNoBundles)
         localRepositoryManager.addQuerySourceFileEclRepository(&instance.queryErrorProcessor(), eclBundlePath.str(), ESFoptional, 0);
-    localRepositoryManager.addQuerySourceFileEclRepository(&instance.queryErrorProcessor(), querySearchPath.str(), 0, 0);
-    instance.dataServer.setown(localRepositoryManager.createCompoundRepository());
 
-    processSingleQuery(instance, NULL, queryAttributePath);
+    if (queryAttributePackage)
+    {
+        instance.dataServer.set(localRepositoryManager.resolveDependentRepository(nullptr, queryAttributePackage, false));
+    }
+    else
+    {
+        const char * searchPath = querySearchPath;
+        while (*searchPath == ENVSEPCHAR)
+            searchPath++;
+
+        if (looksLikeGitPackage(searchPath))
+        {
+            instance.dataServer.set(localRepositoryManager.resolveDependentRepository(nullptr, searchPath, false));
+        }
+        else
+        {
+            localRepositoryManager.addQuerySourceFileEclRepository(&instance.queryErrorProcessor(), searchPath, 0, 0);
+            instance.dataServer.setown(localRepositoryManager.createPackage(nullptr));
+        }
+    }
+
+    processSingleQuery(instance, NULL, queryAttributePath, nullptr);
 
     if (instance.reportErrorSummary())
         return;
@@ -2114,7 +2147,7 @@ bool EclCC::processFiles()
     }
     else if (inputFiles.ordinality() == 0)
     {
-        if (optBatchMode || !optQueryRepositoryReference)
+        if (optBatchMode || !optQueryMainAttribute)
         {
             UERRLOG("No input files could be opened");
             return false;
@@ -2157,9 +2190,8 @@ bool EclCC::processFiles()
     }
     else if (inputFiles.ordinality() == 0)
     {
-        assertex(optQueryRepositoryReference);
         EclCompileInstance info(*this, NULL, *errs, stderr, optOutputFilename, optLegacyImport, optLegacyWhen, optIgnoreSignatures, optIgnoreUnknownImport, optXml);
-        processReference(info, optQueryRepositoryReference);
+        processReference(info, optQueryMainAttribute, optQueryMainPackage);
         ok = (errs->errCount() == 0);
 
         info.logStats(logTimings);
@@ -2768,8 +2800,20 @@ int EclCC::parseCommandLineOptions(int argc, const char* argv[])
         else if (iter.matchOption(optMonitorInterval, "--monitorinterval"))
         {
         }
-        else if (iter.matchOption(optQueryRepositoryReference, "-main"))
+        else if (iter.matchOption(tempArg, "-main"))
         {
+            const char * arg = tempArg;
+            const char * at = strchr(arg, '@');
+            if (at)
+            {
+                optQueryMainAttribute.set(arg, at - arg);
+                optQueryMainPackage.set(at+1);
+            }
+            else
+            {
+                optQueryMainAttribute.set(arg);
+                optQueryMainPackage.set(nullptr);
+            }
         }
         else if (iter.matchFlag(optDebugMemLeak, "-m"))
         {
@@ -2987,7 +3031,7 @@ int EclCC::parseCommandLineOptions(int argc, const char* argv[])
 
     if (inputFileNames.ordinality() == 0 && !optKeywords)
     {
-        if (optGenerateHeader || optShowPaths || (!optBatchMode && optQueryRepositoryReference))
+        if (optGenerateHeader || optShowPaths || (!optBatchMode && optQueryMainAttribute))
             return 0;
         UERRLOG("No input filenames supplied");
         return 1;

--- a/ecl/eclcc/eclcc.hpp
+++ b/ecl/eclcc/eclcc.hpp
@@ -38,7 +38,7 @@ const char * const helpText[] = {
     "                  executable, or stdout)",
     "?   -manifest     Specify path to manifest file listing resources to add",
     "    -foption[=value] Set an ecl option (#option)",
-    "    -main <ref>   Compile definition <ref> from the source collection",
+    "    -main <ref>[@repo] Compile definition <ref> from the source collection, or specified git repo",
     "    -syntax       Perform a syntax check of the ECL",
     "    -platform=hthor Generate code for hthor executable",
     "    -platform=roxie Generate code for roxie cluster (default)",

--- a/ecl/hql/hql.hpp
+++ b/ecl/hql/hql.hpp
@@ -197,6 +197,12 @@ interface IEclRepositoryCallback : public IEclRepository
     virtual IEclSource * getSource(IEclSource * parent, IIdAtom * searchName) = 0;
 };
 
+interface IEclPackage: public IEclRepository
+{
+    virtual const char * queryPackageName() = 0;
+};
+
+
 interface ICodegenContextCallback : public IInterface
 {
     virtual void noteCluster(const char *clusterName) = 0;

--- a/ecl/hql/hqlcache.cpp
+++ b/ecl/hql/hqlcache.cpp
@@ -425,7 +425,7 @@ void ArchiveCreator::createArchiveItem(const char * fullName, IEclSource * origi
         StringBuffer moduleName;
         const char * attrName = splitFullname(moduleName, fullName);
 
-        IPropertyTree * module = queryEnsureArchiveModule(archive, moduleName, nullptr);
+        IPropertyTree * module = queryEnsureArchiveModule(archive, nullptr, moduleName, nullptr);
         assertex(!queryArchiveAttribute(module, attrName));
         IPropertyTree * attr = createArchiveAttribute(module, attrName);
         setDefinitionText(attr, "", original->queryFileContents(), false);
@@ -433,7 +433,7 @@ void ArchiveCreator::createArchiveItem(const char * fullName, IEclSource * origi
     else
     {
         Owned<IProperties> properties = original->getProperties();
-        IPropertyTree * module = queryEnsureArchiveModule(archive, fullName, nullptr);
+        IPropertyTree * module = queryEnsureArchiveModule(archive, nullptr, fullName, nullptr);
         IFileContents * contents = original->queryFileContents();
         setDefinitionText(module, "Text", contents, false);
 

--- a/ecl/hql/hqlcollect.cpp
+++ b/ecl/hql/hqlcollect.cpp
@@ -642,6 +642,7 @@ public:
     }
 
     virtual void populateChildren();
+    virtual const char * queryPath() const override { return elemTree->queryProp("@package"); }
     virtual void setTree(IPropertyTree * _elemTree) { elemTree.set(_elemTree); }
 
     CXmlEclElement * find(IIdAtom * searchName) { return static_cast<CXmlEclElement *>(CEclCollection::find(searchName)); }
@@ -651,6 +652,7 @@ public:
 protected:
     void expandAttribute(const char * modname, IPropertyTree * tree);
     void expandChildren(IPropertyTree * xml);
+    void expandDependency(const char * modname, const char * package, IPropertyTree * tree);
     void expandModule(const char * modname, IPropertyTree * tree);
     void getFullName(StringBuffer & target);
 
@@ -691,7 +693,12 @@ void CXmlEclElement::expandChildren(IPropertyTree * xml)
     {
         IPropertyTree & cur = modit->query();
         const char* modname = cur.queryProp("@name");
-        expandModule(modname, &cur);
+        const char* package = cur.queryProp("@package");
+
+        if (!package)
+            expandModule(modname, &cur);
+        else
+            expandDependency(modname, package, &cur);
     }
 
     Owned<IPropertyTreeIterator> attrit = xml->getElements("Attribute");
@@ -744,6 +751,21 @@ void CXmlEclElement::expandModule(const char * modname, IPropertyTree * tree)
         expandChildren(tree);
 }
 
+
+void CXmlEclElement::expandDependency(const char * modname, const char * package, IPropertyTree * tree)
+{
+    assertex(modname && package);
+    const char * dot = strchr(modname, '.');
+    if (dot)
+    {
+        IIdAtom * name = createIdAtom(modname, dot-modname);
+        select(name, ESTcontainer, NULL)->expandModule(dot+1, tree);
+        return;
+    }
+
+    IIdAtom * thisName = createIdAtom(modname);
+    select(thisName, ESTdependency, tree);
+}
 
 void CXmlEclElement::expandAttribute(const char * modname, IPropertyTree * tree)
 {

--- a/ecl/hql/hqlesp.cpp
+++ b/ecl/hql/hqlesp.cpp
@@ -185,7 +185,7 @@ void WebServicesExtractor::getAttributeText(StringBuffer & text, const char* att
     if(!dot || !dot[1])
         throw MakeStringException(3, "Please specify both module and attribute");
 
-    OwnedHqlExpr symbol = getResolveAttributeFullPath(attributeName, LSFpublic, lookupCtx); 
+    OwnedHqlExpr symbol = getResolveAttributeFullPath(attributeName, LSFpublic, lookupCtx, nullptr);
     if (!symbol || !hasNamedSymbol(symbol) || !symbol->hasText()) 
     {
         StringBuffer txt;
@@ -199,8 +199,9 @@ void WebServicesExtractor::getAttributeText(StringBuffer & text, const char* att
     /* MORE: It would be preferable if this was better integrated with hqlgram2.cpp.  It's a reasonable stopgap */
     if (archive)
     {
+        const char * package = nullptr;     //MORE: GCH->AF not sure how this fits in...
         StringAttr moduleName(attributeName, dot-attributeName);
-        IPropertyTree * moduleTree = queryEnsureArchiveModule(archive, moduleName, NULL);
+        IPropertyTree * moduleTree = queryEnsureArchiveModule(archive, package, moduleName, nullptr);
         IPropertyTree * attrTree = queryArchiveAttribute(moduleTree, dot+1);
         if (!attrTree)
         {
@@ -364,7 +365,7 @@ bool retrieveWebServicesInfo(IWorkUnit *workunit, HqlLookupContext & ctx)
     }
 
     //Names are currently stored case insensitively, we want the case sensitive variant.
-    OwnedHqlExpr s1 = ctx.queryRepository()->queryRootScope()->lookupSymbol(createIdAtom(moduleName.str()), LSFpublic, ctx);
+    OwnedHqlExpr s1 = ctx.queryPackage()->queryRootScope()->lookupSymbol(createIdAtom(moduleName.str()), LSFpublic, ctx);
     if (s1 && s1->queryScope())
     {
         const char *caseSensitiveModuleName = s1->queryScope()->queryFullName();

--- a/ecl/hql/hqlexpr.cpp
+++ b/ecl/hql/hqlexpr.cpp
@@ -908,9 +908,9 @@ void HqlParseContext::addForwardReference(IHqlScope * owner, IHasUnlinkedOwnerRe
     forwardLinks.append(*new ForwardScopeItem(owner, child));
 }
 
-IPropertyTree * HqlParseContext::queryEnsureArchiveModule(const char * name, IHqlScope * scope)
+IPropertyTree * HqlParseContext::queryEnsureArchiveModule(const char * package, const char * name, IHqlScope * scope)
 {
-    return ::queryEnsureArchiveModule(archive, name, scope);
+    return ::queryEnsureArchiveModule(archive, package, name, scope);
 }
 
 void HqlParseContext::setGatherMeta(const MetaOptions & options)
@@ -938,7 +938,7 @@ IPropertyTree * HqlParseContext::beginMetaSource(IFileContents * contents)
     return attr;
 }
 
-void HqlParseContext::noteBeginAttribute(IHqlScope * scope, IFileContents * contents, IIdAtom * name)
+void HqlParseContext::noteBeginAttribute(const char * package, IHqlScope * scope, IFileContents * contents, IIdAtom * name)
 {
     beginMetaScope();
     if (queryNestedDependTree())
@@ -948,7 +948,7 @@ void HqlParseContext::noteBeginAttribute(IHqlScope * scope, IFileContents * cont
     {
         const char * moduleName = scope->queryFullName();
 
-        IPropertyTree * module = queryEnsureArchiveModule(moduleName, scope);
+        IPropertyTree * module = queryEnsureArchiveModule(package, moduleName, scope);
         IPropertyTree * attr = queryArchiveAttribute(module, str(name));
         if (!attr)
             attr = createArchiveAttribute(module, str(name));
@@ -984,7 +984,7 @@ void HqlParseContext::noteBeginQuery(IHqlScope * scope, IFileContents * contents
         const char * moduleName = scope->queryFullName();
         if (moduleName && *moduleName)
         {
-            IPropertyTree * module = queryEnsureArchiveModule(moduleName, scope);
+            IPropertyTree * module = queryEnsureArchiveModule(nullptr, moduleName, scope);
             setDefinitionText(module, "Text", contents, checkDirty);
         }
     }
@@ -999,7 +999,7 @@ void HqlParseContext::noteBeginQuery(IHqlScope * scope, IFileContents * contents
     }
 }
 
-void HqlParseContext::noteBeginModule(IHqlScope * scope, IFileContents * contents)
+void HqlParseContext::noteBeginModule(const char * package, IHqlScope * scope, IFileContents * contents)
 {
     beginMetaScope();
     if (queryNestedDependTree())
@@ -1010,7 +1010,7 @@ void HqlParseContext::noteBeginModule(IHqlScope * scope, IFileContents * content
         const char * moduleName = scope->queryFullName();
         if (moduleName && *moduleName)
         {
-            IPropertyTree * module = queryEnsureArchiveModule(moduleName, scope);
+            IPropertyTree * module = queryEnsureArchiveModule(package, moduleName, scope);
             setDefinitionText(module, "Text", contents, checkDirty);
         }
     }
@@ -1187,7 +1187,7 @@ bool HqlParseContext::includeInArchive(IHqlScope * scope) const
     return archive && (!scope || scope->includeInArchive());
 }
 
-void HqlParseContext::noteExternalLookup(IHqlScope * parentScope, IHqlExpression * expr)
+void HqlParseContext::noteExternalLookup(const char * package, IHqlScope * parentScope, IIdAtom * id, IHqlExpression * expr)
 {
     //metaStack can be empty if we are resolving the main attribute within the repository
     if (!metaStack)
@@ -1200,7 +1200,13 @@ void HqlParseContext::noteExternalLookup(IHqlScope * parentScope, IHqlExpression
         //It would be preferrable to only check once, but adds very little time anyway.
         IHqlScope * resolvedScope = expr->queryScope();
         if (includeInArchive(resolvedScope))
-            queryEnsureArchiveModule(resolvedScope->queryFullName(), resolvedScope);
+        {
+            const char * fullname =  resolvedScope->queryFullName();
+            if (fullname)
+                queryEnsureArchiveModule(package, fullname, resolvedScope);
+            else
+                ensureArchiveImport(archive, package, id, resolvedScope);
+        }
     }
 
     FileParseMeta & meta = metaStack.tos();
@@ -1305,8 +1311,28 @@ extern HQL_API IPropertyTree * createAttributeArchive()
     return archive.getClear();
 }
 
-IPropertyTree * queryEnsureArchiveModule(IPropertyTree * archive, const char * name, IHqlScope * scope)
+static IPropertyTree * querySelectPackage(IPropertyTree * archive, const char * package)
 {
+    if (package)
+    {
+        VStringBuffer xpath("Archive[@package='%s']", package);
+        IPropertyTree * nested = archive->queryPropTree(xpath);
+        if (!nested)
+        {
+            nested = archive->addPropTree("Archive");
+            nested->setProp("@package", package);
+        }
+        archive = nested;
+    }
+    return archive;
+}
+
+//This function ensures that there is an iniitailised <Module> tag within the appropriate <Archive>
+IPropertyTree * queryEnsureArchiveModule(IPropertyTree * archive, const char * package, const char * name, IHqlScope * scope)
+{
+    archive = querySelectPackage(archive, package);
+
+    //Note: name can be null if this is a source file in the root of the repository
     //MORE: Move this into a member of the parse context to also handle dependencies.
     StringBuffer lowerName;
     lowerName.append(name).toLowerCase();
@@ -1339,6 +1365,24 @@ IPropertyTree * queryEnsureArchiveModule(IPropertyTree * archive, const char * n
         }
     }
     return module;
+}
+
+void ensureArchiveImport(IPropertyTree * archive, const char * package, IIdAtom * id, IHqlScope * scope)
+{
+    archive = querySelectPackage(archive, package);
+
+    const char * name = str(id);
+    const char * lowerName = str(lower(id));
+    StringBuffer xpath,s;
+    xpath.append("Module[@key=\"").append(lowerName).append("\"]");
+    IPropertyTree * module = archive->queryPropTree(xpath);
+    if (!module)
+    {
+        module = archive->addPropTree("Module");
+        module->setProp("@name", name);
+        module->setProp("@key", lowerName);
+        module->setProp("@package", scope->queryPackageName());
+    }
 }
 
 extern HQL_API IPropertyTree * queryArchiveAttribute(IPropertyTree * module, const char * name)
@@ -1395,7 +1439,7 @@ HqlLookupContext::~HqlLookupContext()
 
 void HqlLookupContext::noteBeginAttribute(IHqlScope * scope, IFileContents * contents, IIdAtom * name)
 {
-    parseCtx.noteBeginAttribute(scope, contents, name);
+    parseCtx.noteBeginAttribute(rootPackage->queryPackageName(), scope, contents, name);
 }
 
 
@@ -1407,7 +1451,7 @@ void HqlLookupContext::noteBeginQuery(IHqlScope * scope, IFileContents * content
 
 void HqlLookupContext::noteBeginModule(IHqlScope * scope, IFileContents * contents)
 {
-    parseCtx.noteBeginModule(scope, contents);
+    parseCtx.noteBeginModule(rootPackage->queryPackageName(), scope, contents);
 }
 
 
@@ -8751,12 +8795,12 @@ IHqlExpression *CHqlRemoteScope::clone(HqlExprArray &newkids)
 }
 
 
-void CHqlRemoteScope::noteExternalLookup(HqlLookupContext & ctx, IHqlExpression * expr)
+void CHqlRemoteScope::noteExternalLookup(HqlLookupContext & ctx, IIdAtom * searchId, IHqlExpression * expr)
 {
     if (!text)
-        ctx.noteExternalLookup(this, expr);
+        ctx.noteExternalLookup(this, searchId, expr);
     else
-        ctx.noteExternalLookup(nullptr, this);
+        ctx.noteExternalLookup(nullptr, searchId, this);
 }
 
 IHqlExpression *CHqlRemoteScope::lookupSymbol(IIdAtom * searchName, unsigned lookupFlags, HqlLookupContext & ctx)
@@ -8776,7 +8820,7 @@ IHqlExpression *CHqlRemoteScope::lookupSymbol(IIdAtom * searchName, unsigned loo
         if (resolvedSym)
         {
             if (!(lookupFlags & LSFnoreport))
-                noteExternalLookup(ctx, resolvedSym);
+                noteExternalLookup(ctx, searchName, resolvedSym);
             return resolvedSym.getClear();
         }
     }
@@ -8802,7 +8846,7 @@ IHqlExpression *CHqlRemoteScope::lookupSymbol(IIdAtom * searchName, unsigned loo
     if ((lookupFlags & LSFignoreBase))
     {
         if (!(lookupFlags & LSFnoreport))
-            noteExternalLookup(ctx, ret);
+            noteExternalLookup(ctx, searchName, ret);
         return ret.getClear();
     }
 
@@ -8864,7 +8908,7 @@ IHqlExpression *CHqlRemoteScope::lookupSymbol(IIdAtom * searchName, unsigned loo
         return NULL;
 
     if (!(lookupFlags & LSFnoreport))
-        noteExternalLookup(ctx, newSymbol);
+        noteExternalLookup(ctx, searchName, newSymbol);
     return newSymbol.getClear();
 }
 
@@ -9100,10 +9144,10 @@ inline bool canMergeDefinition(IHqlExpression * expr)
 
 IHqlExpression * CHqlMergedScope::lookupSymbol(IIdAtom * searchId, unsigned lookupFlags, HqlLookupContext & ctx)
 {
-    if (rootRepository != ctx.queryRepository())
+    if (rootRepository != ctx.queryPackage())
     {
         HqlLookupContext childCtx(ctx);
-        childCtx.rootRepository = rootRepository;
+        childCtx.rootPackage = rootRepository;
         return lookupSymbol(searchId, lookupFlags, childCtx);
     }
     HqlCriticalBlock block(cs);
@@ -9122,7 +9166,7 @@ IHqlExpression * CHqlMergedScope::lookupSymbol(IIdAtom * searchId, unsigned look
         if (resolvedOp != no_nobody)
         {
             //Resolving from a merged scope, means that the resolved item will be a global symbol
-            ctx.noteExternalLookup(this, resolved);
+            ctx.noteExternalLookup(this, searchId, resolved);
             return resolved.getClear();
         }
     }
@@ -9295,6 +9339,11 @@ bool CHqlMergedScope::isRemoteScope() const
     return false;
 }
 
+
+const char * CHqlMergedScope::queryPackageName() const
+{
+    return rootRepository->queryPackageName();
+}
 
 //==============================================================================================================
 
@@ -16975,7 +17024,7 @@ static void gatherAttributeDependencies(HqlLookupContext & ctx, const char * ite
         else
             moduleName = createIdAtom(item);
 
-        OwnedHqlExpr resolved = ctx.queryRepository()->queryRootScope()->lookupSymbol(moduleName, LSFpublic, ctx);
+        OwnedHqlExpr resolved = ctx.queryPackage()->queryRootScope()->lookupSymbol(moduleName, LSFpublic, ctx);
         if (resolved)
         {
             IHqlScope * scope = resolved->queryScope();
@@ -16994,7 +17043,7 @@ static void gatherAttributeDependencies(HqlLookupContext & ctx, const char * ite
     }
 }
 
-extern HQL_API IPropertyTree * gatherAttributeDependencies(IEclRepository * dataServer, const char * items)
+extern HQL_API IPropertyTree * gatherAttributeDependencies(IEclPackage * dataServer, const char * items)
 {
     NullStatisticTarget nullStats;
     HqlParseContext parseCtx(nullptr, nullptr, nullStats);

--- a/ecl/hql/hqlexpr.hpp
+++ b/ecl/hql/hqlexpr.hpp
@@ -883,8 +883,8 @@ public:
     }
 
     void addForwardReference(IHqlScope * owner, IHasUnlinkedOwnerReference * child);
-    void noteBeginAttribute(IHqlScope * scope, IFileContents * contents, IIdAtom * name);
-    void noteBeginModule(IHqlScope * scope, IFileContents * contents);
+    void noteBeginAttribute(const char * package, IHqlScope * scope, IFileContents * contents, IIdAtom * name);
+    void noteBeginModule(const char * package, IHqlScope * scope, IFileContents * contents);
     void noteBeginQuery(IHqlScope * scope, IFileContents * contents);
     void noteBeginMacro(IHqlScope * scope, IIdAtom * name);
     void noteEndAttribute(bool success);
@@ -893,9 +893,9 @@ public:
     void noteFinishedParse(IHqlScope * scope);
     void noteEndMacro();
     void notePrivateSymbols(IHqlScope * scope);
-    IPropertyTree * queryEnsureArchiveModule(const char * name, IHqlScope * scope);
+    IPropertyTree * queryEnsureArchiveModule(const char * package, const char * name, IHqlScope * scope);
 
-    void noteExternalLookup(IHqlScope * parentScope, IHqlExpression * expr);
+    void noteExternalLookup(const char * package, IHqlScope * parentScope, IIdAtom * name, IHqlExpression * expr);
 
     void setGatherMeta(const MetaOptions & options);
     void setCacheLocation(const char * path);
@@ -986,10 +986,10 @@ public:
         container = &other;
         if (parseCtx.timeParser)
             startCycles = get_cycles_now();
-        rootRepository = other.rootRepository;
+        rootPackage = other.rootPackage;
     }
-    HqlLookupContext(HqlParseContext & _parseCtx, IErrorReceiver * _errs, IEclRepository * _rootRepository)
-    : parseCtx(_parseCtx), errs(_errs), rootRepository(_rootRepository)
+    HqlLookupContext(HqlParseContext & _parseCtx, IErrorReceiver * _errs, IEclPackage * _rootRepository)
+    : parseCtx(_parseCtx), errs(_errs), rootPackage(_rootRepository)
     {
         functionCache = &parseCtx.defaultFunctionCache;
         if (parseCtx.timeParser)
@@ -1008,10 +1008,13 @@ public:
     inline void noteEndMacro() { parseCtx.noteEndMacro(); }
 
     inline void noteFinishedParse(IHqlScope * scope) { parseCtx.noteFinishedParse(scope); }
-    inline void noteExternalLookup(IHqlScope * parentScope, IHqlExpression * expr) { parseCtx.noteExternalLookup(parentScope, expr); }
+    inline void noteExternalLookup(IHqlScope * parentScope, IIdAtom * id, IHqlExpression * expr)
+    {
+        parseCtx.noteExternalLookup(rootPackage->queryPackageName(), parentScope, id, expr);
+    }
     inline void notePrivateSymbols(IHqlScope * scope) { parseCtx.notePrivateSymbols(scope); }
 
-    inline IEclRepository * queryRepository() const { return rootRepository; }
+    inline IEclPackage * queryPackage() const { return rootPackage; }
     inline bool queryExpandCallsWhenBound() const { return parseCtx.expandCallsWhenBound; }
     inline HqlParseContext & queryParseContext() const { return parseCtx; }
     inline unsigned numErrors() const { return errs ? errs->errCount() : 0; }
@@ -1039,7 +1042,7 @@ private:
 public:
     Linked<IErrorReceiver> errs;
     HqlExprArray * functionCache;
-    IEclRepository * rootRepository;
+    IEclPackage * rootPackage;
     cycle_t startCycles = 0;
     cycle_t childCycles = 0;
     HqlLookupContext * container = nullptr;
@@ -1107,6 +1110,7 @@ interface IHqlScope : public IInterface
     virtual bool getProp(IAtom *, StringBuffer &) const = 0;
     virtual bool includeInArchive() const = 0;
     virtual bool isRemoteScope() const = 0;
+    virtual const char * queryPackageName() const = 0;
 
     //IHqlCreateScope
     virtual void defineSymbol(IIdAtom * name, IIdAtom * moduleName, IHqlExpression *value, bool isExported, bool isShared, unsigned flags, IFileContents *fc, int lineno, int column, int _startpos, int _bodypos, int _endpos) = 0;
@@ -1476,7 +1480,7 @@ extern HQL_API IHqlExpression *parseQuery(IHqlScope *scope, IFileContents * cont
                                           HqlLookupContext & ctx, IXmlScope *xmlScope, IProperties * macroParams, bool loadImplicit, bool isRoot);
 extern HQL_API IHqlExpression *parseQuery(const char *in, IErrorReceiver * errs);
 
-extern HQL_API IPropertyTree * gatherAttributeDependencies(IEclRepository * dataServer, const char * items = NULL);
+extern HQL_API IPropertyTree * gatherAttributeDependencies(IEclPackage * dataServer, const char * items = NULL);
 
 extern HQL_API IHqlScope *createService();
 extern HQL_API IHqlScope *createDatabase(IHqlExpression * name);

--- a/ecl/hql/hqlexpr.ipp
+++ b/ecl/hql/hqlexpr.ipp
@@ -1166,6 +1166,7 @@ public:
     virtual bool getProp(IAtom * a, StringBuffer &ret) const override { return false; }
     virtual bool includeInArchive() const override { return false; }
     virtual bool isRemoteScope() const override { return false; }
+    virtual const char * queryPackageName() const override { return nullptr; }
 
     using CHqlDelayedCall::clone;
     virtual IHqlScope * clone(HqlExprArray & children, HqlExprArray & symbols) override { throwUnexpected(); }
@@ -1231,6 +1232,7 @@ public:
     virtual bool includeInArchive() const override { return true; }
     virtual bool isContainerScope() const override { return false; }
     virtual bool isRemoteScope() const override { return false; }
+    virtual const char * queryPackageName() const override { return nullptr; }
 
     virtual void    getSymbols(HqlExprArray& exprs) const override;
     virtual IHqlScope * clone(HqlExprArray & children, HqlExprArray & symbols) override { throwUnexpected(); }
@@ -1295,7 +1297,7 @@ protected:
     void doParseScopeText(HqlLookupContext & ctx);
     IHqlExpression * repositoryLoadSymbol(IIdAtom * attrName);
     void repositoryLoadModule(HqlLookupContext & ctx, bool forceAll);
-    void noteExternalLookup(HqlLookupContext & ctx, IHqlExpression * expr);
+    void noteExternalLookup(HqlLookupContext & ctx, IIdAtom * searchId, IHqlExpression * expr);
 
     virtual bool equals(const IHqlExpression & other) const override;
 
@@ -1333,6 +1335,7 @@ public:
     virtual bool includeInArchive() const override;
     virtual IEclSource * queryEclSource() const override { return eclSource; }
     virtual bool isRemoteScope() const override { return true; }
+    virtual const char * queryPackageName() const override { throwUnexpected(); }
 };
 
 class HQL_API CHqlLocalScope : public CHqlScope
@@ -1392,7 +1395,7 @@ public:
 class HQL_API CHqlMergedScope : public CHqlScope
 {
 public:
-    CHqlMergedScope(IIdAtom * _id, const char * _fullName, CHqlMergedScope * _parent, IEclRepository * _rootRepository)
+    CHqlMergedScope(IIdAtom * _id, const char * _fullName, CHqlMergedScope * _parent, IEclPackage * _rootRepository)
      : CHqlScope(no_mergedscope, _id, _fullName), parent(_parent), rootRepository(_rootRepository)
     {
     }
@@ -1408,12 +1411,13 @@ public:
     virtual bool isContainerScope() const override { return true; }
     virtual IHqlScope * queryConcreteScope() override { return this; }
     virtual bool isRemoteScope() const override;
+    virtual const char * queryPackageName() const override;
 
 protected:
     CriticalSection cs;
     HqlScopeArray mergedScopes;
     CHqlMergedScope * parent;
-    IEclRepository * rootRepository;
+    IEclPackage * rootRepository;
     bool mergedAll = false;
 };
 
@@ -1647,6 +1651,7 @@ public:
     virtual bool getProp(IAtom *, StringBuffer &) const override { return false; }
     virtual bool includeInArchive() const override { return false; }
     virtual bool isRemoteScope() const override { return false; }
+    virtual const char * queryPackageName() const override { return nullptr; }
 
 //IHqlCreateScope
     virtual void defineSymbol(IIdAtom * id, IIdAtom * moduleName, IHqlExpression *value, bool isExported, bool isShared, unsigned flags, IFileContents *fc, int lineno, int column, int _startpos, int _bodypos, int _endpos) override { throwUnexpected(); }

--- a/ecl/hql/hqlgram2.cpp
+++ b/ecl/hql/hqlgram2.cpp
@@ -360,8 +360,8 @@ HqlGram::HqlGram(IHqlScope * _globalScope, IHqlScope * _containerScope, IFileCon
     lexObject->setLegacyWhen(queryLegacyWhenSemantics());
 
     //MORE: This should be in the parseContext calculated once
-    if (lookupCtx.queryRepository() && loadImplicit)
-        getImplicitScopes(implicitScopes, lookupCtx.queryRepository(), _containerScope, lookupCtx);
+    if (lookupCtx.queryPackage() && loadImplicit)
+        getImplicitScopes(implicitScopes, lookupCtx.queryPackage(), _containerScope, lookupCtx);
 }
 
 
@@ -455,7 +455,7 @@ IHqlScope * HqlGram::queryMacroScope()
     const char * scopeName = lexObject->queryMacroScopeName();
     if (scopeName)
     {
-        OwnedHqlExpr matched = getResolveAttributeFullPath(scopeName, LSFpublic, lookupCtx);
+        OwnedHqlExpr matched = getResolveAttributeFullPath(scopeName, LSFpublic, lookupCtx, nullptr);
         if (matched && matched->queryScope())
             return matched->queryScope();
     }
@@ -11061,12 +11061,12 @@ IHqlExpression * HqlGram::resolveImportModule(const attribute & errpos, IHqlExpr
     if (isHashDollarModule(expr))
         return LINK(queryExpression(queryMacroScope()));
     if (isRootModule(expr))
-        return LINK(queryExpression(lookupCtx.queryRepository()->queryRootScope()));
+        return LINK(queryExpression(lookupCtx.queryPackage()->queryRootScope()));
 
     IAtom * name = expr->queryName();
     if ((name != _dot_Atom) && (name != _container_Atom))
     {
-        if (!lookupCtx.queryRepository())
+        if (!lookupCtx.queryPackage())
         {
             //This never happens in practice since a null repository is generally passed.
             reportError(ERR_MODULE_UNKNOWN, "Import not supported with no repository specified",  
@@ -11077,7 +11077,8 @@ IHqlExpression * HqlGram::resolveImportModule(const attribute & errpos, IHqlExpr
         }
 
         IIdAtom * id = expr->queryId();
-        OwnedHqlExpr importMatch = lookupCtx.queryRepository()->queryRootScope()->lookupSymbol(id, LSFimport, lookupCtx);
+        IHqlScope * rootScope = lookupCtx.queryPackage()->queryRootScope();
+        OwnedHqlExpr importMatch = rootScope->lookupSymbol(id, LSFimport, lookupCtx);
         if (!importMatch)
             importMatch.setown(lookupParseSymbol(id));
 
@@ -11114,7 +11115,7 @@ IHqlExpression * HqlGram::resolveImportModule(const attribute & errpos, IHqlExpr
         //scope as the parent, rather than the merged scope...
         if (containerName)
         {
-            OwnedHqlExpr matched = getResolveAttributeFullPath(containerName, LSFpublic, lookupCtx);
+            OwnedHqlExpr matched = getResolveAttributeFullPath(containerName, LSFpublic, lookupCtx, nullptr);
             if (matched)
                 return matched.getClear();
         }
@@ -11122,7 +11123,7 @@ IHqlExpression * HqlGram::resolveImportModule(const attribute & errpos, IHqlExpr
         {
             //A null parent must be because it is within the global scope, or is the global scope
             if (parentName)
-                return LINK(lookupCtx.queryRepository()->queryRootScope()->queryExpression());
+                return LINK(lookupCtx.queryPackage()->queryRootScope()->queryExpression());
         }
 
         if (parentName)

--- a/ecl/hql/hqlparse.cpp
+++ b/ecl/hql/hqlparse.cpp
@@ -1396,7 +1396,7 @@ void HqlLex::doInModule(attribute & returnToken)
 
 static bool isInModule(HqlLookupContext & ctx, const char* moduleName, const char* attrName)
 {
-    if (!ctx.queryRepository())
+    if (!ctx.queryPackage())
         return false;
 
     try
@@ -1408,7 +1408,7 @@ static bool isInModule(HqlLookupContext & ctx, const char* moduleName, const cha
         const char* pAttr = attrName;
         while(*pAttr==' ') pAttr++;
 
-        OwnedHqlExpr match = ctx.queryRepository()->queryRootScope()->lookupSymbol(createIdAtom(pModule), LSFpublic, ctx);
+        OwnedHqlExpr match = ctx.queryPackage()->queryRootScope()->lookupSymbol(createIdAtom(pModule), LSFpublic, ctx);
         IHqlScope * scope = match ? match->queryScope() : NULL;
         if (scope)
         {

--- a/ecl/hql/hqlplugininfo.cpp
+++ b/ecl/hql/hqlplugininfo.cpp
@@ -26,7 +26,7 @@
 
 namespace repositoryCommon {
 
-IEclRepository * loadPlugins(const char * pluginPath)
+IEclPackage * loadPlugins(const char * pluginPath)
 {
     MultiErrorReceiver errs;
     EclRepositoryManager collection;
@@ -36,7 +36,7 @@ IEclRepository * loadPlugins(const char * pluginPath)
         StringBuffer s;
         DBGLOG(0,"Errors in plugins: %s", errs.toString(s).str());
     }
-    return collection.createCompoundRepository();
+    return collection.createPackage(nullptr);
 }
 
 static void expandPluginPropertyTree(IPropertyTree * target, HqlLookupContext & ctx, IHqlScope * scope, bool includeModuleText)
@@ -85,7 +85,7 @@ static void expandPluginPropertyTree(IPropertyTree * target, HqlLookupContext & 
     }
 }
 
-IPropertyTree * createPluginPropertyTree(IEclRepository * plugins, bool includeModuleText)
+IPropertyTree * createPluginPropertyTree(IEclPackage * plugins, bool includeModuleText)
 {
     NullStatisticTarget nullStats;
     HqlParseContext parseCtx(nullptr, nullptr, nullStats);
@@ -98,7 +98,7 @@ IPropertyTree * createPluginPropertyTree(IEclRepository * plugins, bool includeM
 
 //-------------------------------------------------------------------------------------------------
 
-IPropertyTree * getPlugin(IPropertyTree * p, IEclRepository * plugins, const char* modname, bool load)
+IPropertyTree * getPlugin(IPropertyTree * p, IEclPackage * plugins, const char* modname, bool load)
 {
     if (!p)
         return NULL;

--- a/ecl/hql/hqlplugininfo.hpp
+++ b/ecl/hql/hqlplugininfo.hpp
@@ -22,9 +22,9 @@
 
 namespace repositoryCommon {
 
-extern HQL_API IEclRepository * loadPlugins(const char * pluginPath);
-extern HQL_API IPropertyTree * createPluginPropertyTree(IEclRepository * plugins, bool includeModuleText);
-extern HQL_API IPropertyTree * getPlugin(IPropertyTree * p, IEclRepository * plugins, const char* modname, bool load);
+extern HQL_API IEclPackage * loadPlugins(const char * pluginPath);
+extern HQL_API IPropertyTree * createPluginPropertyTree(IEclPackage * plugins, bool includeModuleText);
+extern HQL_API IPropertyTree * getPlugin(IPropertyTree * p, IEclPackage * plugins, const char* modname, bool load);
 
 }   //namespace
 

--- a/ecl/hql/hqlrepository.cpp
+++ b/ecl/hql/hqlrepository.cpp
@@ -39,9 +39,16 @@ static const char * queryExtractFilename(const char * urn)
     return nullptr;
 }
 
-bool looksLikeGitUrl(const char * urn)
+bool startsWithGitProtocol(const char * urn)
 {
     if (startsWith(urn, "git+") || startsWith(urn, "git:"))
+        return true;
+    return false;
+}
+
+bool looksLikeGitPackage(const char * urn)
+{
+    if (startsWithGitProtocol(urn))
         return true;
     if (strstr(urn, ".git#"))
         return true;
@@ -52,7 +59,7 @@ static bool splitRepoVersion(StringBuffer & repoUrn, StringBuffer & repo, String
 {
     const char * cur = urn;
     //Allow either protocol://<server>/<user>/<repo>[#version] or <user>/<repo>[#version]
-    if (startsWith(urn, "git+") || startsWith(urn, "git:"))
+    if (startsWithGitProtocol(urn))
     {
         //skip to the end of the protocol
         const char * colon = strchr(urn, ':');
@@ -144,7 +151,7 @@ void getImplicitScopes(HqlScopeArray& implicitScopes, IEclRepository * repositor
 
 extern HQL_API void importRootModulesToScope(IHqlScope * scope, HqlLookupContext & ctx)
 {
-    IEclRepository * eclRepository = ctx.queryRepository();
+    IEclRepository * eclRepository = ctx.queryPackage();
     if (eclRepository)
     {
         HqlExprArray rootSymbols;
@@ -164,9 +171,10 @@ extern HQL_API void importRootModulesToScope(IHqlScope * scope, HqlLookupContext
 
 //-------------------------------------------------------------------------------------------------------------------
 
-IHqlExpression * getResolveAttributeFullPath(const char * attrname, unsigned lookupFlags, HqlLookupContext & ctx)
+IHqlExpression * getResolveAttributeFullPath(const char * attrname, unsigned lookupFlags, HqlLookupContext & ctx, IEclPackage * optPackage)
 {
-    Linked<IHqlScope> parentScope = ctx.queryRepository()->queryRootScope();
+    IEclPackage * package = optPackage ? optPackage : ctx.queryPackage();
+    Linked<IHqlScope> parentScope = package->queryRootScope();
     const char * item = attrname;
     for (;;)
     {
@@ -202,8 +210,8 @@ IHqlExpression * getResolveAttributeFullPath(const char * attrname, unsigned loo
 IHqlScope * getResolveDottedScope(const char * modname, unsigned lookupFlags, HqlLookupContext & ctx)
 {
     if (!modname || !*modname)
-        return LINK(ctx.queryRepository()->queryRootScope());
-    OwnedHqlExpr matched = getResolveAttributeFullPath(modname, lookupFlags, ctx);
+        return LINK(ctx.queryPackage()->queryRootScope());
+    OwnedHqlExpr matched = getResolveAttributeFullPath(modname, lookupFlags, ctx, nullptr);
     if (matched)
         return LINK(matched->queryScope());
     return NULL;
@@ -212,20 +220,26 @@ IHqlScope * getResolveDottedScope(const char * modname, unsigned lookupFlags, Hq
 
 //-------------------------------------------------------------------------------------------------------------------
 
-class HQL_API CompoundEclRepository : implements CInterfaceOf<IEclRepository>
+class HQL_API CompoundEclRepository : implements CInterfaceOf<IEclPackage>
 {
 public:
-    CompoundEclRepository() { rootScope.setown(new CHqlMergedScope(nullptr, nullptr, nullptr, this)); }
+    CompoundEclRepository(const char * _packageName)
+    : packageName(_packageName)
+    {
+        rootScope.setown(new CHqlMergedScope(nullptr, nullptr, nullptr, this));
+    }
 
     void addRepository(IEclRepository & _repository);
 
     virtual IHqlScope * queryRootScope() { return rootScope; }
     virtual IEclSource * getSource(const char * eclFullname) override;
     virtual bool includeInArchive() const override { return true; }
+    virtual const char * queryPackageName() { return packageName; }
 
 protected:
     IArrayOf<IEclRepository> repositories;
     Owned<CHqlMergedScope> rootScope;
+    StringAttr packageName;
 };
 
 void CompoundEclRepository::addRepository(IEclRepository & _repository)
@@ -248,9 +262,9 @@ IEclSource * CompoundEclRepository::getSource(const char * eclFullname)
 
 //-------------------------------------------------------------------------------------------------------------------
 
-extern HQL_API IEclRepository * createCompoundRepositoryF(IEclRepository * repository, ...)
+extern HQL_API IEclPackage * createCompoundRepositoryF(const char * packageName, IEclRepository * repository, ...)
 {
-    Owned<CompoundEclRepository> compound = new CompoundEclRepository;
+    Owned<CompoundEclRepository> compound = new CompoundEclRepository(packageName);
     compound->addRepository(*repository);
     va_list args;
     va_start(args, repository);
@@ -266,9 +280,9 @@ extern HQL_API IEclRepository * createCompoundRepositoryF(IEclRepository * repos
 }
 
 
-extern HQL_API IEclRepository * createCompoundRepository(EclRepositoryArray & repositories)
+extern HQL_API IEclPackage * createCompoundRepository(const char * packageName, EclRepositoryArray & repositories)
 {
-    Owned<CompoundEclRepository> compound = new CompoundEclRepository;
+    Owned<CompoundEclRepository> compound = new CompoundEclRepository(packageName);
     ForEachItemIn(i, repositories)
         compound->addRepository(repositories.item(i));
     return compound.getClear();
@@ -535,7 +549,6 @@ void EclRepositoryManager::addRepository(IEclSourceCollection * source, const ch
     allSources.append(*repo.getClear());
 }
 
-
 void EclRepositoryManager::addMapping(const char * url, const char * path)
 {
     StringBuffer repoUrn, repo, version;
@@ -544,7 +557,31 @@ void EclRepositoryManager::addMapping(const char * url, const char * path)
     repos.append(*new EclRepositoryMapping(repo, version, path));
 }
 
-IEclRepository * EclRepositoryManager::resolveDependentRepository(IIdAtom * name, const char * defaultUrl, bool requireSHA)
+void EclRepositoryManager::processArchive(IPropertyTree * archiveTree)
+{
+    IArrayOf<IEclRepository> savedSources;        // also includes -D options
+    savedSources.swapWith(allSources);
+
+    Owned<IPropertyTreeIterator> subArchives = archiveTree->getElements("Archive");
+    ForEach(*subArchives)
+    {
+        IPropertyTree & cur = subArchives->query();
+        const char * defaultUrl = cur.queryProp("@package");
+        const char * repoKey = defaultUrl;
+        Owned<IEclSourceCollection> archiveCollection(createArchiveEclCollection(&cur));
+        ForEachItemIn(iShared, sharedSources)
+            allSources.append(OLINK(sharedSources.item(iShared)));
+        allSources.append(*createRepository(archiveCollection, nullptr, true));
+        Owned<IEclPackage> compound = createPackage(defaultUrl);
+        dependencies.emplace_back(repoKey, compound);
+    }
+
+    allSources.swapWith(savedSources);
+    Owned<IEclSourceCollection> archiveCollection(createArchiveEclCollection(archiveTree));
+    addRepository(archiveCollection, nullptr, true);
+}
+
+IEclPackage * EclRepositoryManager::resolveDependentRepository(IIdAtom * name, const char * defaultUrl, bool requireSHA)
 {
     //Check to see if the reference is to a filename.  Should possibly be disabled on a switch.
     const char * filename = queryExtractFilename(defaultUrl);
@@ -659,7 +696,7 @@ IEclRepository * EclRepositoryManager::resolveDependentRepository(IIdAtom * name
     unsigned flags = 0;
     Owned<IEclRepository> repo = createNewSourceFileEclRepository(errs, filename, flags, 0, true);
     allSources.append(*repo.getClear());
-    Owned<IEclRepository> compound = createCompoundRepository();
+    Owned<IEclPackage> compound = createPackage(defaultUrl);
     dependencies.emplace_back(repoKey, compound);
     return compound;
 }
@@ -667,14 +704,6 @@ IEclRepository * EclRepositoryManager::resolveDependentRepository(IIdAtom * name
 
 IEclRepository * EclRepositoryManager::createNewSourceFileEclRepository(IErrorReceiver *errs, const char * path, unsigned flags, unsigned trace, bool includeInArchive)
 {
-    if (*path == ENVSEPCHAR)
-        path++;
-
-    if (looksLikeGitUrl(path))
-    {
-        IEclRepository * match = EclRepositoryManager::resolveDependentRepository(nullptr, path, false);
-        return LINK(match);
-    }
     Owned<IEclSourceCollection> source = createFileSystemEclCollection(errs, path, flags, trace);
     return createRepository(source, nullptr, includeInArchive);
 }
@@ -685,9 +714,9 @@ IEclRepository * EclRepositoryManager::createSingleDefinitionEclRepository(const
     return createRepository(source, nullptr, includeInArchive);
 }
 
-IEclRepository * EclRepositoryManager::createCompoundRepository()
+IEclPackage * EclRepositoryManager::createPackage(const char * packageName)
 {
-    Owned<IEclRepository> compound = ::createCompoundRepository(allSources);
+    Owned<IEclPackage> compound = ::createCompoundRepository(packageName, allSources);
     dependencies.emplace_back("", compound);
     allSources.kill();
     return compound.getClear();

--- a/ecl/hql/hqlrepository.hpp
+++ b/ecl/hql/hqlrepository.hpp
@@ -49,11 +49,12 @@ public:
 
     void addMapping(const char * url, const char * path);
 
-    IEclRepository * createCompoundRepository();
+    IEclPackage * createPackage(const char * packageName);
     void inherit(const EclRepositoryManager & other);
     void kill();
 
-    IEclRepository * resolveDependentRepository(IIdAtom * name, const char * defaultUrl, bool requireSHA);
+    void processArchive(IPropertyTree * archiveTree);
+    IEclPackage * resolveDependentRepository(IIdAtom * name, const char * defaultUrl, bool requireSHA);
     void setOptions(const char * _eclRepoPath, bool _fetchRepos, bool _updateRepos, bool _verbose)
     {
         options.eclRepoPath.set(_eclRepoPath);
@@ -70,7 +71,7 @@ protected:
     unsigned runGitCommand(StringBuffer * output, const char *args, const char * cwd);
 
 private:
-    using DependencyInfo = std::pair<std::string, Shared<IEclRepository>>;
+    using DependencyInfo = std::pair<std::string, Shared<IEclPackage>>;
     CIArrayOf<EclRepositoryMapping> repos;
     std::vector<DependencyInfo> dependencies;
     IArrayOf<IEclRepository> sharedSources;     // plugins, std library, bundles
@@ -86,8 +87,8 @@ private:
 };
 
 
-extern HQL_API IEclRepository * createCompoundRepositoryF(IEclRepository * repository, ...);
-extern HQL_API IEclRepository * createCompoundRepository(EclRepositoryArray & repositories);
+extern HQL_API IEclPackage * createCompoundRepositoryF(const char * packageName, IEclRepository * repository, ...);
+extern HQL_API IEclPackage * createCompoundRepository(const char * packageName, EclRepositoryArray & repositories);
 extern HQL_API IEclRepository * createNestedRepository(IIdAtom * name, IEclRepository * root);
 
 extern HQL_API void getRootScopes(HqlScopeArray & rootScopes, IHqlScope * scope, HqlLookupContext & ctx);
@@ -96,6 +97,7 @@ extern HQL_API void getImplicitScopes(HqlScopeArray & implicitScopes, IEclReposi
 extern HQL_API void importRootModulesToScope(IHqlScope * scope, HqlLookupContext & ctx);
 
 extern HQL_API IHqlScope * getResolveDottedScope(const char * modname, unsigned lookupFlags, HqlLookupContext & ctx);
-extern HQL_API IHqlExpression * getResolveAttributeFullPath(const char * attrname, unsigned lookupFlags, HqlLookupContext & ctx);
+extern HQL_API IHqlExpression * getResolveAttributeFullPath(const char * attrname, unsigned lookupFlags, HqlLookupContext & ctx, IEclPackage * optPackage);
+extern HQL_API bool looksLikeGitPackage(const char * urn);
 
 #endif

--- a/ecl/hql/hqlutil.hpp
+++ b/ecl/hql/hqlutil.hpp
@@ -755,10 +755,11 @@ extern HQL_API IHqlExpression * normalizeDatasetAlias(IHqlExpression * expr);
 extern HQL_API IHqlExpression * normalizeAnyDatasetAliases(IHqlExpression * expr);
 
 //In hqlgram2.cpp
-extern HQL_API IPropertyTree * queryEnsureArchiveModule(IPropertyTree * archive, const char * name, IHqlScope * rScope);
+extern HQL_API IPropertyTree * queryEnsureArchiveModule(IPropertyTree * archive, const char * package, const char * name, IHqlScope * rScope);
 extern HQL_API IPropertyTree * queryArchiveAttribute(IPropertyTree * module, const char * name);
 extern HQL_API IPropertyTree * createArchiveAttribute(IPropertyTree * module, const char * name);
 extern HQL_API IPropertyTree * queryArchiveEntry(IPropertyTree * archive, const char * name);
+extern HQL_API void ensureArchiveImport(IPropertyTree * archive, const char * package, IIdAtom * id, IHqlScope * scope);
 
 extern HQL_API IError * annotateExceptionWithLocation(IException * e, IHqlExpression * location);
 extern HQL_API IHqlExpression * expandMacroDefinition(IHqlExpression * expr, HqlLookupContext & ctx, bool reportError);

